File: web/backend/services/indexer.js
"use strict";

const fs = require("fs");
const path = require("path");
const {
  resolveWorkspaceDir,
  resolveAssetsDir,
  safeReadJson,
  safeReadYaml,
  listDirectories,
} = require("./workspace-reader");

const DEFAULT_THRESHOLDS = {
  DR: 0.95,
  CPR: 0.9,
  Reward: 0.7,
};

/**
 * Atomic write: write to temp file then rename.
 * @param {string} targetPath
 * @param {string} content
 */
function atomicWrite(targetPath, content) {
  const dir = path.dirname(targetPath);
  fs.mkdirSync(dir, { recursive: true });
  const tmpPath = `${targetPath}.tmp`;
  fs.writeFileSync(tmpPath, content, "utf-8");
  fs.renameSync(tmpPath, targetPath);
}

/**
 * Build runs_index.json by scanning workspace/runs.
 * @param {string} workspaceDir
 * @returns {{runs: any[]}}
 */
function buildRunsIndex(workspaceDir) {
  const runsDir = path.join(workspaceDir, "runs");
  const runs = [];

  if (fs.existsSync(runsDir)) {
    const runIds = listDirectories(runsDir).sort().reverse();
    for (const runId of runIds) {
      const runDir = path.join(runsDir, runId);
      const summaryPath = path.join(runDir, "reports", "summary.json");
      const manifestPath = path.join(runDir, "run_manifest.json");

      const summary = safeReadJson(summaryPath);
      if (!summary) continue;

      const manifest = safeReadJson(manifestPath, {});
      const entry = {
        run_id: summary.run_id || runId,
        mode: manifest.mode || "unknown",
        total_samples: summary.total_samples || 0,
        metrics: summary.metrics || {},
        failure_breakdown: summary.failure_breakdown || {},
        created_at: manifest.created_at || "",
        project: manifest.project || null,
      };
      runs.push(entry);
    }
  }

  return { runs };
}

/**
 * Load project configs from assets/projects/*.yaml.
 * @param {string} assetsDir
 * @returns {{projects: any[]}}
 */
function loadProjects(assetsDir) {
  const projectsDir = path.join(assetsDir, "projects");
  const projects = [];

  if (fs.existsSync(projectsDir)) {
    const files = fs
      .readdirSync(projectsDir)
      .filter((f) => f.endsWith(".yaml") || f.endsWith(".yml"))
      .sort();
    for (const file of files) {
      const filePath = path.join(projectsDir, file);
      const raw = safeReadYaml(filePath, {});
      const data = raw.project || raw;
      const projectId = data.id || path.basename(file, path.extname(file));
      projects.push({
        id: projectId,
        name: data.name || projectId,
        description: data.description || "",
        default_rule_set: data.default_rule_set || "",
        default_task_set: data.default_task_set || "",
        created_at: data.created_at || "",
        latest_run_id: null,
        run_count: 0,
      });
    }
  }

  return { projects };
}

/**
 * Cross-reference projects with runs to compute run_count and latest_run_id.
 * @param {{projects: any[]}} projectsIndex
 * @param {{runs: any[]}} runsIndex
 */
function crossReferenceProjects(projectsIndex, runsIndex) {
  const projMap = new Map(projectsIndex.projects.map((p) => [p.id, p]));
  for (const run of runsIndex.runs) {
    const projectId = run.project;
    if (projectId && projMap.has(projectId)) {
      const proj = projMap.get(projectId);
      proj.run_count += 1;
      if (!proj.latest_run_id) {
        proj.latest_run_id = run.run_id;
      }
    }
  }
}

/**
 * Rebuild workspace index files.
 * @param {string} [workspaceDir]
 * @param {string} [assetsDir]
 * @returns {{run_count: number, project_count: number}}
 */
function rebuild(workspaceDir, assetsDir) {
  const ws = workspaceDir || resolveWorkspaceDir();
  const assets = assetsDir || resolveAssetsDir();
  const indexDir = path.join(ws, "index");

  const runsIndex = buildRunsIndex(ws);
  const projectsIndex = loadProjects(assets);
  crossReferenceProjects(projectsIndex, runsIndex);

  atomicWrite(
    path.join(indexDir, "runs_index.json"),
    JSON.stringify(runsIndex, null, 2)
  );
  atomicWrite(
    path.join(indexDir, "projects.json"),
    JSON.stringify(projectsIndex, null, 2)
  );

  return {
    run_count: runsIndex.runs.length,
    project_count: projectsIndex.projects.length,
  };
}

module.exports = {
  rebuild,
  buildRunsIndex,
  loadProjects,
  crossReferenceProjects,
  DEFAULT_THRESHOLDS,
};

File: evaluator/agent_eval/web/indexer.py
"""Workspace 索引重建 — 供 Web Portal 与 CLI `agent-eval index` 使用。"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

import yaml


def _atomic_write_json(path: Path, data: dict[str, Any]) -> None:
    """原子写入 JSON 文件。"""
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.with_suffix(path.suffix + ".tmp")
    tmp.write_text(json.dumps(data, ensure_ascii=False, indent=2), encoding="utf-8")
    tmp.replace(path)


def _load_project_configs(assets_dir: Path) -> list[dict[str, Any]]:
    """从 assets/projects/*.yaml 加载项目配置。"""
    projects_dir = assets_dir / "projects"
    projects: list[dict[str, Any]] = []
    if not projects_dir.exists():
        return projects

    for file in sorted(projects_dir.glob("*.yaml")):
        try:
            raw = yaml.safe_load(file.read_text(encoding="utf-8")) or {}
            data = raw.get("project", raw)
            project_id = data.get("id") or file.stem
            projects.append(
                {
                    "id": project_id,
                    "name": data.get("name", project_id),
                    "description": data.get("description", ""),
                    "default_rule_set": data.get("default_rule_set", ""),
                    "default_task_set": data.get("default_task_set", ""),
                    "created_at": data.get("created_at", ""),
                    "latest_run_id": None,
                    "run_count": 0,
                }
            )
        except Exception:
            # 跳过无效项目配置
            continue
    return projects


def _scan_runs(runs_dir: Path) -> list[dict[str, Any]]:
    """扫描 workspace/runs 目录，重建 runs_index。"""
    runs: list[dict[str, Any]] = []
    if not runs_dir.exists():
        return runs

    for run_dir in sorted(runs_dir.iterdir(), reverse=True):
        if not run_dir.is_dir():
            continue
        summary_path = run_dir / "reports" / "summary.json"
        manifest_path = run_dir / "run_manifest.json"
        if not summary_path.exists():
            continue

        try:
            summary = json.loads(summary_path.read_text(encoding="utf-8"))
            manifest = {}
            if manifest_path.exists():
                manifest = json.loads(manifest_path.read_text(encoding="utf-8"))

            runs.append(
                {
                    "run_id": summary.get("run_id", run_dir.name),
                    "mode": manifest.get("mode", "unknown"),
                    "total_samples": summary.get("total_samples", 0),
                    "metrics": summary.get("metrics", {}),
                    "failure_breakdown": summary.get("failure_breakdown", {}),
                    "created_at": manifest.get("created_at", ""),
                    "project": manifest.get("project"),
                }
            )
        except Exception:
            continue

    return runs


def _cross_reference_projects(projects: list[dict[str, Any]], runs: list[dict[str, Any]]) -> None:
    """根据 runs 计算每个项目的 run_count 和 latest_run_id。"""
    proj_map = {p["id"]: p for p in projects}
    for run in runs:
        project_id = run.get("project")
        if project_id and project_id in proj_map:
            proj = proj_map[project_id]
            proj["run_count"] += 1
            if proj["latest_run_id"] is None:
                proj["latest_run_id"] = run["run_id"]


def rebuild_index(workspace_dir: Path | str) -> dict[str, Any]:
    """重建 workspace 索引。

    生成 workspace/index/projects.json 与 workspace/index/runs_index.json。

    Args:
        workspace_dir: workspace 根目录。

    Returns:
        统计信息 dict：{"project_count": int, "run_count": int}
    """
    workspace = Path(workspace_dir).resolve()
    index_dir = workspace / "index"
    runs_dir = workspace / "runs"
    assets_dir = workspace.parent / "assets"

    runs = _scan_runs(runs_dir)
    projects = _load_project_configs(assets_dir)
    _cross_reference_projects(projects, runs)

    _atomic_write_json(index_dir / "runs_index.json", {"runs": runs})
    _atomic_write_json(index_dir / "projects.json", {"projects": projects})

    return {
        "project_count": len(projects),
        "run_count": len(runs),
    }
